lnwire/onion_error.go

```diff
@@ -0,0 +1,879 @@
+package lnwire
+
+import (
+	"crypto/sha256"
+	"io"
+
+	"bytes"
+
+	"github.com/go-errors/errors"
+	"github.com/roasbeef/btcutil"
+)
+
+// Failure represent the onion failure object which is able to identify itself
+// by returning the unique failure code.
+type Failure interface {
+	Code() FailCode
+}
+
+// fixedFailureMessageLength is the size of the failure message plus the size
+// of padding. Failure message always should be equals to this size.
+const fixedFailureMessageLength = 128
+
+const (
+	// FlagBadOnion error flag denotes unparsable onion, encrypted by
+	// previous node.
+	FlagBadOnion FailCode = 0x8000
+
+	// FlagPerm error flag denotes permanent failure.
+	FlagPerm FailCode = 0x4000
+
+	// FlagNode error flag denotes node failure.
+	FlagNode FailCode = 0x2000
+
+	// FlagUpdate error flag denotes new channel update enclosed.
+	FlagUpdate FailCode = 0x1000
+)
+
+// FailCode specifies the precise reason that an upstream HTLC was cancelled.
+// Each UpdateFailHTLC message carries a FailCode which is to be passed back
+// unaltered to the source of the HTLC within the route.
+type FailCode uint16
+
+const (
+	CodeInvalidRealm                  FailCode = FlagBadOnion | 1
```

Comment:
I have added only one comment on top of the `const` block, as you did it with `lnwire` messages types.